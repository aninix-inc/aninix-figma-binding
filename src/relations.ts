export class Relations {
  private _map = new Map<string, Set<string>>()

  public addRelation = (from: string, to: string): this => {
    let set = this._map.get(from)

    if (set === undefined) {
      set = new Set<string>()
      this._map.set(from, set)
    }

    set.add(to)

    return this
  }

  toJSON = (): Record<string, string[]> => {
    const map: Record<string, string[]> = {}

    for (const key of this._map.keys()) {
      const set = this._map.get(key)

      if (set === undefined) {
        continue
      }

      map[key] = Array.from(set.values())
    }

    return map
  }
}
