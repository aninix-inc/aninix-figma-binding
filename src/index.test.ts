import { beforeEach, describe, expect, mock, test } from 'bun:test'
import {
  getNodeId,
  mapEntityBaseProperties,
  mapEntityBlendProperties,
  mapEntityCornerProperties,
  mapEntityEntryProperties,
  mapEntityFrameProperties,
  mapEntityInstanceProperties,
  mapEntitySceneProperties,
} from '.'
import { Relations } from './relations'

// @NOTE: provide global figma object
// @ts-ignore
globalThis.figma = {
  mixed: Symbol('figma.mixed'),
}

describe('mapEntityEntryProperties', () => {
  test('with parent', () => {
    const result = mapEntityEntryProperties({
      projectId: 'test-project',
      nodeId: 'test-node',
      parentNodeId: 'hello',
    })
    expect(result).toMatchSnapshot()
  })

  test('without parent', () => {
    const result = mapEntityEntryProperties({
      projectId: 'test-project',
      nodeId: 'test-node',
    })
    expect(result).toMatchSnapshot()
  })
})

describe('mapEntityBaseProperties', () => {
  let relations: Relations

  beforeEach(() => {
    relations = new Relations()
  })

  test('has NOT parent', () => {
    const result = mapEntityBaseProperties(
      relations,
      {
        name: 'Some Node',
        type: 'ELLIPSE',
      },
      {
        projectId: 'test-project',
        nodeId: 'test-node',
      }
    )
    expect(result).toMatchSnapshot()
    expect(relations.toJSON()).toMatchSnapshot()
  })

  test('has parent', () => {
    const result = mapEntityBaseProperties(
      relations,
      {
        name: 'Some Node',
        type: 'ELLIPSE',
      },
      {
        projectId: 'test-project',
        nodeId: 'test-node',
        parentNodeId: 'test-parent-node',
      }
    )
    expect(result).toMatchSnapshot()
    expect(relations.toJSON()).toMatchSnapshot()
  })

  test('has NOT initial node', () => {
    const result = mapEntityBaseProperties(
      relations,
      {
        name: 'Some Node',
        type: 'ELLIPSE',
      },
      {
        projectId: 'test-project',
        nodeId: 'test-node',
      }
    )
    expect(result).toMatchSnapshot()
    expect(relations.toJSON()).toMatchSnapshot()
  })

  test('has initial node', () => {
    const result = mapEntityBaseProperties(
      relations,
      {
        name: 'Some Node',
        type: 'ELLIPSE',
      },
      {
        projectId: 'test-project',
        nodeId: 'test-node',
        initialNodeId: 'initial-test-node',
        parentNodeId: 'test-parent-node',
      }
    )
    expect(result).toMatchSnapshot()
    expect(relations.toJSON()).toMatchSnapshot()
  })
})

test('mapEntitySceneProperties', () => {
  const result = mapEntitySceneProperties({
    locked: true,
    visible: true,
  })
  expect(result).toMatchSnapshot()
})

describe('mapEntityBlendProperties', () => {
  let relations: Relations

  beforeEach(() => {
    relations = new Relations()
  })

  test('has NOT effects', () => {
    const entities: any[] = []
    const result = mapEntityBlendProperties(
      entities,
      relations,
      {
        blendMode: 'NORMAL',
        isMask: false,
        effects: [],
      },
      {
        projectId: 'test-project',
        nodeId: 'test-node',
        initialNodeId: 'initial-test-node',
        parentNodeId: 'test-parent-node',
      }
    )
    expect(entities).toMatchSnapshot()
    expect(relations.toJSON()).toMatchSnapshot()
    expect(result).toMatchSnapshot()
  })

  test('has drop shadow', () => {
    const entities: any[] = []
    const result = mapEntityBlendProperties(
      entities,
      relations,
      {
        blendMode: 'NORMAL',
        isMask: false,
        effects: [
          {
            type: 'DROP_SHADOW',
            color: {
              r: 255,
              g: 0,
              b: 0,
              a: 1,
            },
            offset: {
              x: 0,
              y: 0,
            },
            radius: 10,
            visible: true,
            blendMode: 'NORMAL',
          },
        ],
      },
      {
        projectId: 'test-project',
        nodeId: 'test-node',
        initialNodeId: 'initial-test-node',
        parentNodeId: 'test-parent-node',
      }
    )
    expect(entities).toMatchSnapshot()
    expect(relations.toJSON()).toMatchSnapshot()
    expect(result).toMatchSnapshot()
  })

  test('has inner shadow', () => {
    const entities: any[] = []
    const result = mapEntityBlendProperties(
      entities,
      relations,
      {
        blendMode: 'NORMAL',
        isMask: false,
        effects: [
          {
            type: 'INNER_SHADOW',
            color: {
              r: 255,
              g: 0,
              b: 0,
              a: 1,
            },
            offset: {
              x: 0,
              y: 0,
            },
            radius: 10,
            visible: true,
            blendMode: 'NORMAL',
          },
        ],
      },
      {
        projectId: 'test-project',
        nodeId: 'test-node',
        initialNodeId: 'initial-test-node',
        parentNodeId: 'test-parent-node',
      }
    )
    expect(entities).toMatchSnapshot()
    expect(relations.toJSON()).toMatchSnapshot()
    expect(result).toMatchSnapshot()
  })

  test('has layer blur', () => {
    const entities: any[] = []
    const result = mapEntityBlendProperties(
      entities,
      relations,
      {
        blendMode: 'NORMAL',
        isMask: false,
        effects: [
          {
            type: 'LAYER_BLUR',
            radius: 10,
            visible: true,
          },
        ],
      },
      {
        projectId: 'test-project',
        nodeId: 'test-node',
        initialNodeId: 'initial-test-node',
        parentNodeId: 'test-parent-node',
      }
    )
    expect(entities).toMatchSnapshot()
    expect(relations.toJSON()).toMatchSnapshot()
    expect(result).toMatchSnapshot()
  })

  test('has background blur', () => {
    const entities: any[] = []
    const result = mapEntityBlendProperties(
      entities,
      relations,
      {
        blendMode: 'NORMAL',
        isMask: false,
        effects: [
          {
            type: 'BACKGROUND_BLUR',
            radius: 10,
            visible: true,
          },
        ],
      },
      {
        projectId: 'test-project',
        nodeId: 'test-node',
        initialNodeId: 'initial-test-node',
        parentNodeId: 'test-parent-node',
      }
    )
    expect(entities).toMatchSnapshot()
    expect(relations.toJSON()).toMatchSnapshot()
    expect(result).toMatchSnapshot()
  })

  test('has multiple effects', () => {
    const entities: any[] = []
    const result = mapEntityBlendProperties(
      entities,
      relations,
      {
        blendMode: 'NORMAL',
        isMask: false,
        effects: [
          {
            type: 'DROP_SHADOW',
            color: {
              r: 255,
              g: 0,
              b: 0,
              a: 1,
            },
            offset: {
              x: 0,
              y: 0,
            },
            radius: 10,
            visible: true,
            blendMode: 'NORMAL',
          },
          {
            type: 'INNER_SHADOW',
            color: {
              r: 255,
              g: 0,
              b: 0,
              a: 1,
            },
            offset: {
              x: 0,
              y: 0,
            },
            radius: 10,
            visible: true,
            blendMode: 'NORMAL',
          },
          {
            type: 'LAYER_BLUR',
            radius: 10,
            visible: true,
          },
          {
            type: 'BACKGROUND_BLUR',
            radius: 10,
            visible: true,
          },
        ],
      },
      {
        projectId: 'test-project',
        nodeId: 'test-node',
        initialNodeId: 'initial-test-node',
        parentNodeId: 'test-parent-node',
      }
    )
    expect(entities).toMatchSnapshot()
    expect(relations.toJSON()).toMatchSnapshot()
    expect(result).toMatchSnapshot()
  })
})

test('mapEntityCornerProperties', () => {
  const result = mapEntityCornerProperties({
    cornerRadius: 15,
    cornerSmoothing: 0.5,
  })
  expect(result).toMatchSnapshot()
})

test('mapEntityFrameProperties', () => {
  const result = mapEntityFrameProperties({
    clipsContent: true,
  })
  expect(result).toMatchSnapshot()
})

describe('mapEntityInstanceProperties', () => {
  test('not null', () => {
    const getPluginData = mock()
    const getSharedPluginData = mock().mockReturnValue('some-id-from-aninix')

    const result = mapEntityInstanceProperties(
      [],
      {
        id: 'some-id-1-from-figma',
        name: 'some-name-1',
        getPluginData,
        getSharedPluginData,
        mainComponent: {
          id: 'some-id-2-from-figma',
          name: 'some-name-2',
          getPluginData,
          getSharedPluginData,
        },
      },
      getNodeId
    )
    expect(result).toMatchSnapshot()
  })

  test('null', () => {
    const getPluginData = mock()
    const getSharedPluginData = mock().mockReturnValue('some-id-from-aninix')

    expect(() =>
      mapEntityInstanceProperties(
        [],
        {
          id: 'some-id-1-from-figma',
          name: 'some-name-1',
          getPluginData,
          getSharedPluginData,
          mainComponent: null,
        },
        getNodeId
      )
    ).toThrow()
  })
})
