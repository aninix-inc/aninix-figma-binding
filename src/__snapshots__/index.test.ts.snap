// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`mapEntityEntryProperties with parent 1`] = `{}`;

exports[`mapEntityEntryProperties without parent 1`] = `
{
  "entry": true,
}
`;

exports[`mapEntityBaseProperties has NOT parent 1`] = `
{
  "entityType": "NODE",
  "name": "Some Node",
  "nodeType": "ELLIPSE",
}
`;

exports[`mapEntityBaseProperties has NOT parent 2`] = `{}`;

exports[`mapEntityBaseProperties has parent 1`] = `
{
  "entityType": "NODE",
  "name": "Some Node",
  "nodeType": "ELLIPSE",
}
`;

exports[`mapEntityBaseProperties has parent 2`] = `
{
  "test-parent-node/children": [
    "test-node/parent",
  ],
}
`;

exports[`mapEntityBaseProperties has NOT initial node 1`] = `
{
  "entityType": "NODE",
  "name": "Some Node",
  "nodeType": "ELLIPSE",
}
`;

exports[`mapEntityBaseProperties has NOT initial node 2`] = `{}`;

exports[`mapEntityBaseProperties has initial node 1`] = `
{
  "entityType": "NODE",
  "initialNodeId": "initial-test-node",
  "name": "Some Node",
  "nodeType": "ELLIPSE",
}
`;

exports[`mapEntityBaseProperties has initial node 2`] = `
{
  "test-parent-node/children": [
    "test-node/parent",
  ],
}
`;

exports[`mapEntityBlendProperties has NOT effects 1`] = `[]`;

exports[`mapEntityBlendProperties has NOT effects 2`] = `{}`;

exports[`mapEntityBlendProperties has NOT effects 3`] = `
{
  "blendMode": "NORMAL",
  "mask": false,
}
`;

exports[`mapEntityBlendProperties has drop shadow 1`] = `
[
  {
    "components": {
      "blendMode": "NORMAL",
      "effectType": "DROP_SHADOW",
      "entityType": "EFFECT",
      "propertiesExpanded": false,
      "shadowColor": [
        65025,
        0,
        0,
        1,
      ],
      "shadowOffset": [
        0,
        0,
      ],
      "shadowRadius": 10,
      "shadowSpread": 0,
      "visibleInViewport": true,
    },
    "id": "test-nodee0",
    "schemaVersion": 1,
    "tag": "dropShadow",
  },
]
`;

exports[`mapEntityBlendProperties has drop shadow 2`] = `
{
  "test-node/effects": [
    "test-nodee0/parent",
  ],
}
`;

exports[`mapEntityBlendProperties has drop shadow 3`] = `
{
  "blendMode": "NORMAL",
  "mask": false,
}
`;

exports[`mapEntityBlendProperties has inner shadow 1`] = `
[
  {
    "components": {
      "blendMode": "NORMAL",
      "effectType": "INNER_SHADOW",
      "entityType": "EFFECT",
      "propertiesExpanded": false,
      "shadowColor": [
        65025,
        0,
        0,
        1,
      ],
      "shadowOffset": [
        0,
        0,
      ],
      "shadowRadius": 10,
      "shadowSpread": 0,
      "visibleInViewport": true,
    },
    "id": "test-nodee0",
    "schemaVersion": 1,
    "tag": "innerShadow",
  },
]
`;

exports[`mapEntityBlendProperties has inner shadow 2`] = `
{
  "test-node/effects": [
    "test-nodee0/parent",
  ],
}
`;

exports[`mapEntityBlendProperties has inner shadow 3`] = `
{
  "blendMode": "NORMAL",
  "mask": false,
}
`;

exports[`mapEntityBlendProperties has layer blur 1`] = `
[
  {
    "components": {
      "blurRadius": 10,
      "effectType": "LAYER_BLUR",
      "entityType": "EFFECT",
      "propertiesExpanded": false,
      "visibleInViewport": true,
    },
    "id": "test-nodee0",
    "schemaVersion": 1,
    "tag": "layerBlur",
  },
]
`;

exports[`mapEntityBlendProperties has layer blur 2`] = `
{
  "test-node/effects": [
    "test-nodee0/parent",
  ],
}
`;

exports[`mapEntityBlendProperties has layer blur 3`] = `
{
  "blendMode": "NORMAL",
  "mask": false,
}
`;

exports[`mapEntityBlendProperties has background blur 1`] = `
[
  {
    "components": {
      "blurRadius": 10,
      "effectType": "BACKGROUND_BLUR",
      "entityType": "EFFECT",
      "propertiesExpanded": false,
      "visibleInViewport": true,
    },
    "id": "test-nodee0",
    "schemaVersion": 1,
    "tag": "backgroundBlur",
  },
]
`;

exports[`mapEntityBlendProperties has background blur 2`] = `
{
  "test-node/effects": [
    "test-nodee0/parent",
  ],
}
`;

exports[`mapEntityBlendProperties has background blur 3`] = `
{
  "blendMode": "NORMAL",
  "mask": false,
}
`;

exports[`mapEntityBlendProperties has multiple effects 1`] = `
[
  {
    "components": {
      "blendMode": "NORMAL",
      "effectType": "DROP_SHADOW",
      "entityType": "EFFECT",
      "propertiesExpanded": false,
      "shadowColor": [
        65025,
        0,
        0,
        1,
      ],
      "shadowOffset": [
        0,
        0,
      ],
      "shadowRadius": 10,
      "shadowSpread": 0,
      "visibleInViewport": true,
    },
    "id": "test-nodee0",
    "schemaVersion": 1,
    "tag": "dropShadow",
  },
  {
    "components": {
      "blendMode": "NORMAL",
      "effectType": "INNER_SHADOW",
      "entityType": "EFFECT",
      "propertiesExpanded": false,
      "shadowColor": [
        65025,
        0,
        0,
        1,
      ],
      "shadowOffset": [
        0,
        0,
      ],
      "shadowRadius": 10,
      "shadowSpread": 0,
      "visibleInViewport": true,
    },
    "id": "test-nodee1",
    "schemaVersion": 1,
    "tag": "innerShadow",
  },
  {
    "components": {
      "blurRadius": 10,
      "effectType": "LAYER_BLUR",
      "entityType": "EFFECT",
      "propertiesExpanded": false,
      "visibleInViewport": true,
    },
    "id": "test-nodee2",
    "schemaVersion": 1,
    "tag": "layerBlur",
  },
  {
    "components": {
      "blurRadius": 10,
      "effectType": "BACKGROUND_BLUR",
      "entityType": "EFFECT",
      "propertiesExpanded": false,
      "visibleInViewport": true,
    },
    "id": "test-nodee3",
    "schemaVersion": 1,
    "tag": "backgroundBlur",
  },
]
`;

exports[`mapEntityBlendProperties has multiple effects 2`] = `
{
  "test-node/effects": [
    "test-nodee0/parent",
    "test-nodee1/parent",
    "test-nodee2/parent",
    "test-nodee3/parent",
  ],
}
`;

exports[`mapEntityBlendProperties has multiple effects 3`] = `
{
  "blendMode": "NORMAL",
  "mask": false,
}
`;

exports[`mapEntityInstanceProperties not null 1`] = `
{
  "mainNodeComponentId": "some-id-from-aninix",
}
`;

exports[`mapEntitySceneProperties 1`] = `
{
  "locked": true,
  "nodeColor": "BLUE",
  "propertiesExpanded": false,
  "solo": "NONE",
  "visibleInViewport": true,
}
`;

exports[`mapEntityCornerProperties 1`] = `
{
  "cornerRadius": 15,
  "smoothCornerRadius": true,
}
`;

exports[`mapEntityFrameProperties 1`] = `
{
  "clipContent": true,
}
`;
